### ISSUE TITLE
ipptransform support

### ISSUE THREAD

**Issue by michaelrsweet** [michaelrsweet/pappl#308]
Add support for using ipptransform (or any other external command) to provide filtering to the driver native format.

**Comment by DemiMarie** [comment]
This should include sandboxing of the external command, together with stripping of any metadata that could be used to exploit the printer.  Specifically, I would use the following approach:

1. In the sandboxed process, convert the input to a raw RGB/YUV bitmap (X by Y pixels, no metadata, X and Y of limited size such that X * Y does not overflow).
2. In the parent process, convert the bitmap to whatever the printer needs.

**Comment by michaelrsweet** [comment]
First, ipptransform doesn't just produce raster data, it also produces PCL, PostScript, and PDF content for printing.

Second, doing a raw bitmap is a totally unnecessary and costly (in terms of time, memory, and filesystem space) step that is more likely to cause failures due to the size of the bitmap data: ~150MB per page for A4/Letter on a typical color printer. Internally the PDF RIP generates strips of (raw) bitmap data, up to the whole page depending on the size, which is then PackBits compressed and streamed as PWG or Apple raster data (the only two raster formats supported by ipptransform...)

**Comment by michaelrsweet** [comment]
Sandboxing resources:

- [Landlock](https://landlock.io): Userland sandboxing/restrictions on recent Linux kernels.
- [Sandbox](https://igorstechnoclub.com/sandbox-exec/): macOS sandboxing support; this isn't an Apple site, as they have removed most of the online documentation for sandbox-exec and friends, but "man sandbox_init" and "man 7 sandbox" are available on a Mac.

The primary sandboxing goal is to limit access to spool files (only the file(s) being filtered), only allowing writes to the temporary directory, and not allowing network connections.

**Comment by DemiMarie** [comment]
> First, ipptransform doesn't just produce raster data, it also produces PCL, PostScript, and PDF content for printing.
> 
> Second, doing a raw bitmap is a totally unnecessary and costly (in terms of time, memory, and filesystem space) step that is more likely to cause failures due to the size of the bitmap data: ~150MB per page for A4/Letter on a typical color printer. Internally the PDF RIP generates strips of (raw) bitmap data, up to the whole page depending on the size, which is then PackBits compressed and streamed as PWG or Apple raster data (the only two raster formats supported by ipptransform...)

I agree that buffering a whole bitmap in memory is a bad idea.  That said, I want to protect agaisst chained attacks: a malicious document exploits `ipptransform` and causes it to produce a malicious raster file, which then exploits the RIP in the printer itself.  Printer firmware is not exactly known for being extremely secure, and the business practices of printer vendors have made users (understandably) reluctant to upgrade firmware in case it introduces anti-features (such as rejecting third-party cartriges).

**Comment by michaelrsweet** [comment]
Well, when it comes to PWG or Apple raster, I made the format as simple as possible, using a CODEC (PackBits) the printer vendors were familiar with. In the grand scheme of things it is probably the safest PDL supported by printers...

Given the number of common security features in use (ASLR, read-only code pages, etc.), the odds of crafting a PDF (or other) file that could lead to code execution or data corruption specific enough to crash a particular printer are so remote they aren't worth the engineering time - there are much easier ways of getting a payload to a printer to crash it.

My concern for using ipptransform (or any external program, really), is limiting what the program is allowed to do to protect the server running the printer application and any information it might have access to.

Some other things I want to do:

- Explicitly control the environment variables that are passed to ipptransform
- Create a per-process TMPDIR that is purged after the run - that would be the only writeable directory
- Include scheduling parameter/policy options when running via `posix_spawn`.
- Run as a non-root user when the printer application is running as root (stupid Linux USB access limitation)

**Comment by DemiMarie** [comment]
> Given the number of common security features in use (ASLR, read-only code pages, etc.), the odds of crafting a PDF (or other) file that could lead to code execution or data corruption specific enough to crash a particular printer are so remote they aren't worth the engineering time - there are much easier ways of getting a payload to a printer to crash it.

This is true of network printers, but it is not true in the cases of:

- USB printers
- Printers on an isolated network segment that only CUPS can access.

I expect many if not most people who care about security will be using the latter if possible.

**Comment by michaelrsweet** [comment]
@DemiMarie Most printer applications also support "raw" printing of printer-ready files (PCL, etc.), so you just submit the raw file and the printer application will "copy" it to the printer, usually without validation/preprocessing. It is probably worth adding a PAPPL system option to disable "raw" printing entirely, but given the number of users that depend on that functionality (particularly for common shipping, retail, and enterprise software) I'm not sure it should be the default, at least for those printer applications it is most used for (LPrint in particular).

All that said, the focus of *this* issue is on supporting the ipptransform command to expand the range of formats that a PAPPL-based printer application can support out of the box, and more specifically to take PWG raster data from this command and pass it through the driver that is embedded in the printer application. Isolating the PDF interpreter from the printer application process is a key security step, as is managing the execution of and resources used by that process and protecting the local system and any locally-accessible data from attack. This will naturally help protect against attacks on the printer itself but we cannot know all the sorts of content that might be dangerous to the printer, particularly when many historical attacks have used entirely valid content otherwise.

**Comment by michaelrsweet** [comment]
Tracking a generic "run command" API in issue #387. The "ipptransform" support will use this API to safely run it...

**Comment by michaelrsweet** [comment]
[master 1c8174e] Save initial work on ipptransform integration (Issue #308)
[master 65cd511] Implement ipptransform raster filter (Issue #308)
